### ISSUE TITLE
Memory leak in ServiceBusClient get_subscription_receiver

### ISSUE THREAD

**Issue by miroag** [Azure/azure-sdk-for-python#19777]
- **Package Name**: azure-servicebus
- **Package Version**: azure-servicebus==7.3.1 && uamqp==1.4.1
- **Operating System**: Windows 10/WSL Ubuntu/Docker python 3.9
- **Python Version**: 3.9

**Describe the bug**
Memory leak in get_subscription_receiver

**To Reproduce**
Snippet below will result in memory leak of few Mb in 10 minutes
```
import os
import time


from azure.servicebus import ServiceBusClient, NEXT_AVAILABLE_SESSION

if __name__ == '__main__':
    sb_client = ServiceBusClient.from_connection_string(os.environ['SB_CONNECTION_STRING'])
    start = time.time()
    i=0
    while time.time() - start < 10 * 60:
        with sb_client.get_subscription_receiver(topic_name='sync', subscription_name='data',
                                                 idle_timeout=1, session_id=NEXT_AVAILABLE_SESSION) as receiver:
            i+=1
            print(f'Receiver #{i}')
```

**Expected behavior**
No memory leaks obviously. The code like in the snippet above results in 1Gb memory leak in 12 hrs in our production environment. Unless, I'm missing something I cannot reuse the receiver in the scenario when I need to work with sessions.
Effectively, this renders the use of ServiceBus very problematic.

**Screenshots**
![image](https://user-images.githubusercontent.com/10284158/125480818-b27bd267-9318-4074-8d64-68fa1ca37fef.png)

**Additional context**
This might be similar / identical to https://github.com/Azure/azure-sdk-for-python/issues/15747
But the above one is closed apparently without proper resolution.

Kind regards,
Alex

**Comment by xiangyan99** [comment]
Thanks for the feedback, we’ll investigate asap.

**Comment by yunhaoling** [comment]
hey @miroag, I'll take a look into the issue this week, thanks for bring it to my attention!

**Comment by yunhaoling** [comment]
hey @miroag , confirmed that memory leak is happening on Windows, also did a thorough debugging and long running test to find the leaking places. I had a [PR](https://github.com/Azure/azure-uamqp-python/pull/258) out for the fix in the uamqp library, but am now waiting for the C team to verify my fix is the right approach.

However, I have run long running test on Linux as well as valgrind (c program memory leak debug tool) but didn't observe any memory leak. Could you help confirm whether there's memory leak on Linux or not?

**Comment by miroag** [comment]
Hello @yunhaoling 

Yes, I can confirm that the issue was observed as well on the AKS k8s Linux (Ubuntu) based and WSL2 Ubuntu.
Below is the image on the WSL2 Ubuntu tests:

![image](https://user-images.githubusercontent.com/10284158/126617423-3f38de6d-34d6-4e5c-bd3b-68809a997a56.png)

Kind regards,
Alex

**Comment by yunhaoling** [comment]
actually I find that you need to tweak the code a bit to properly free the receiver instances which are created from the service bus client.

When a receiver is created from the client, [the client will append the receiver in its private instance variable self._handler](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/azure/servicebus/_servicebus_client.py#L460) such that when a client is closed, every sender/receiver related to the client will be closed.

so in the code, the client shall also be closed to free its handlers otherwise it would just keep growing.

You could update the code as following:

```python
import os
import time

from azure.servicebus import ServiceBusClient, NEXT_AVAILABLE_SESSION

if __name__ == '__main__':

    start = time.time()
    i=0
    while time.time() - start < 10 * 60:
        with ServiceBusClient.from_connection_string(os.environ['SB_CONNECTION_STRING']) as sb_client:
            with sb_client.get_subscription_receiver(topic_name='sync', subscription_name='data',
                                                     idle_timeout=1, session_id=NEXT_AVAILABLE_SESSION) as receiver:
                i+=1
                print(f'Receiver #{i}')
```

**Comment by miroag** [comment]
Well .. this is a bit of an issue ... 
The snippet I've created is the closest thing I can write to illustrate the use in production. Fundamentally it's a daemon that listens for new sessions and when one is available start the receiver and process it and the sb_client is never closed.

I would say receiver (when getting out of scope) shall de-register itself. An alternative might be that I need to call a close method explicitly, something like: 
```
receiver.close()
# or
sb_client.close_receiver(receiver)

```

**Comment by miroag** [comment]
btw: consider example code: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/servicebus/azure-servicebus/samples/sync_samples/session_pool_receive.py
According to your explanation, I think, it will leak the memory in exactly the same way

**Comment by yunhaoling** [comment]
The de-registering is a good point, previously we were thinking that as the client takes responsibility for spawning receivers, so it's also client's duty to track those handlers. But given your issue, tracking those closed handlers seems only to increase memory usage while there's no other usage of those closed handler.

I'll see what I could do to improve the situation.

**Comment by yunhaoling** [comment]
The de-registering is a good point, previously we were thinking that as the client takes responsibility for spawning receivers, so it's also client's duty to track those handlers. But given your issue, tracking those closed handlers seems only to increase memory usage while there's no other usage of those closed handler.

I'll see what I could do to improve the situation.

**Comment by miroag** [comment]
@yunhaoling I've changed my production code to recreate `sb_client` after every 1K sessions. I can confirm that memory leak has been reduced, but it did not fix the issue completely. To compare in previous situation 500Mb of mem will be leaked in about 16hrs, with the force recreation it requires about twice the time. 
AKS / Ubuntu / Python 3.9

**Comment by yunhaoling** [comment]
hey @miroag , thanks for the data. I'm testing the program with closing client in each iteration on an azure linux vm now. I should get some data after letting it run for one day.

**Comment by yunhaoling** [comment]
hey @miroag, I have run a test script for over 3 days, but the memory increase is almost neglectable.

```
time, memory usage percent
2021-08-02 21:10:50.414985 0.3932086512771554
2021-08-02 21:10:51.471502 0.44138468316164287
2021-08-02 21:10:52.491469 0.44319986969904207
2021-08-02 21:10:53.513472 0.44324892879464745
2021-08-02 21:10:55.537723 0.44324892879464745
2021-08-02 21:10:56.560284 0.4432979878902528
...
2021-08-04 05:39:58.835417 0.44432822889796597
2021-08-04 05:39:59.854786 0.44432822889796597
...
2021-08-05 03:41:10.325674 0.44437728799357135
2021-08-05 03:41:11.346262 0.44437728799357135
2021-08-05 03:41:12.366736 0.44437728799357135
2021-08-05 03:41:13.386194 0.44437728799357135
...
2021-08-05 15:02:28.165368 0.44437728799357135
2021-08-05 15:02:29.185932 0.44437728799357135
```

could you help use my code for testing?

```python
import os
import gc
import psutil
from datetime import datetime
from azure.servicebus import ServiceBusClient, NEXT_AVAILABLE_SESSION


CONNECTION_STR = os.environ['SERVICE_BUS_CONNECTION_STR']
TOPIC_NAME = os.environ["SERVICE_BUS_TOPIC_NAME"]
SUBSCRIPTION_NAME = os.environ["SERVICE_BUS_SUBSCRIPTION_NAME"]
process_instance = psutil.Process(os.getpid())
servicebus_client = ServiceBusClient.from_connection_string(conn_str=CONNECTION_STR)
i = 0
while True:
    try:
        with servicebus_client:
            receiver = servicebus_client.get_subscription_receiver(
                topic_name=TOPIC_NAME,
                subscription_name=SUBSCRIPTION_NAME,
                session_id=NEXT_AVAILABLE_SESSION
            )
            print(datetime.now(), process_instance.memory_percent())
            with receiver:
                pass
            i += 1
            gc.collect()
    except:
        print('error encountered')

```

**Comment by yunhaoling** [comment]
hey @miroag , no worries, I will try creating receiver + receive one msg in the while loop in my test scripts.

**Comment by yunhaoling** [comment]
hey @miroag, I could reproduce the memory leak by the scenario of creating receiver + receive one msg.

I've used valgrind to located the issue which is caused by not freeing up a C string when converting an AMQPValue object into a string and already got a PR out to fix it: https://github.com/Azure/azure-uamqp-python/pull/269.

I will also merge my windows mem leak fix PR: https://github.com/Azure/azure-uamqp-python/pull/258.

So should be able to do a release by the end of this month/early next month for the mem leak fix.

Thanks for all the help you have provided and the patience with me.

**Comment by miroag** [comment]
This is a great news, looking forward to give it a try.

**Comment by yunhaoling** [comment]
hey @miroag , we have released uamqp v1.4.2 which fixed the two memory leak issue.

Please try installing the latest uamqp via command `pip install uamqp --grade`.
Let me know if you're still observing memory leaks.
